ui/components/app/signature-request/signature-request-message/signature-request-message.js

```diff
@@ -0,0 +1,168 @@
+} from '../../../../../shared/modules/hexstring-utils';
+import Box from '../../../ui/box';
+import Typography from '../../../ui/typography';
+import {
+  DISPLAY,
+  ALIGN_ITEMS,
+  JUSTIFY_CONTENT,
+  COLORS,
+  FONT_WEIGHT,
+  FLEX_DIRECTION,
+  TYPOGRAPHY,
+  SIZES,
+} from '../../../../helpers/constants/design-system';
+
+export default function SignatureRequestMessage({
+  data,
+  onMessageScrolled,
+  setMessageRootRef,
+  messageRootRef,
+  messageIsScrollable,
+}) {
+  const t = useContext(I18nContext);
+  const [messageIsScrolled, setMessageIsScrolled] = useState(false);
+  const identities = useSelector(getMetaMaskIdentities);
+  const setMessageIsScrolledAtBottom = () => {
+    if (!messageRootRef || messageIsScrolled) {
+      return;
+    }
+
+    const { scrollTop, offsetHeight, scrollHeight } = messageRootRef;
+    const isAtBottom = Math.round(scrollTop) + offsetHeight >= scrollHeight;
+
+    if (isAtBottom) {
+      setMessageIsScrolled(true);
+      onMessageScrolled();
+    }
+  };
+
+  const renderNode = (renderData) => {
+    return (
+      <Box className="signature-request-message__node">
+        {Object.entries(renderData).map(([label, value], i) => (
+          <Box
+            className="signature-request-message__node"
+            key={i}
+            paddingLeft={2}
+            display={
+              typeof value !== 'object' || value === null ? DISPLAY.FLEX : null
+            }
+          >
+            <Typography
+              as="span"
+              color={COLORS.TEXT_DEFAULT}
+              marginLeft={4}
+              fontWeight={
+                typeof value === 'object'
+                  ? FONT_WEIGHT.BOLD
+                  : FONT_WEIGHT.NORMAL
+              }
+            >
+              {label.charAt(0).toUpperCase() + label.slice(1)}:{' '}
+            </Typography>
+            {typeof value === 'object' && value !== null ? (
+              renderNode(value)
+            ) : (
+              <Typography
+                as="span"
+                color={COLORS.TEXT_DEFAULT}
+                marginLeft={4}
+                className="signature-request-message__node__value"
+              >
+                {isValidHexAddress(value, {
+                  mixedCaseUseChecksum: true,
+                }) ? (
+                  <Typography
+                    variant={TYPOGRAPHY.H7}
+                    color={COLORS.INFO_DEFAULT}
+                    className="signature-request-message__node__value__address"
+                  >
+                    <Address
+                      addressOnly
+                      checksummedRecipientAddress={toChecksumHexAddress(value)}
+                      recipientName={getAccountName(identities, value)}
+                    />
+                  </Typography>
+                ) : (
+                  `${value}`
+                )}
+              </Typography>
+            )}
+          </Box>
+        ))}
+      </Box>
+    );
+  };
```

Comment:
We shouldn't create these functions for intermediary components, just render them inline unless they are rendered multiple times then lift them into a new component.

Reply:
@brad-decker Done [here](https://github.com/MetaMask/metamask-extension/pull/16018/commits/a89af984b68e0bb7bfb7a65ebdc34d4876113eb9).